src/Microsoft.AspNetCore.Razor.Language/Legacy/TokenizerBackedParser.cs

```diff
@@ -570,108 +492,202 @@ protected SyntaxToken AcceptWhiteSpaceInLines()
                     lastWs = null;
                 }
 
-                _tokenizer.Next();
+                NextToken();
             }
+
             return lastWs;
         }
 
-        protected bool AtIdentifier(bool allowKeywords)
+        protected internal bool Optional(SyntaxKind type)
         {
-            return CurrentToken != null &&
-                   (Language.IsIdentifier(CurrentToken) ||
-                    (allowKeywords && Language.IsKeyword(CurrentToken)));
+            if (At(type))
+            {
+                AcceptAndMoveNext();
+                return true;
+            }
+            return false;
         }
 
-        // Don't open this to sub classes because it's lazy but it looks eager.
-        // You have to advance the Enumerable to read the next characters.
-        internal IEnumerable<SyntaxToken> ReadWhileLazy(Func<SyntaxToken, bool> condition)
+        protected internal bool Balance(SyntaxListBuilder<RazorSyntaxNode> builder, BalancingModes mode)
         {
-            while (EnsureCurrent() && condition(CurrentToken))
+            var left = CurrentToken.Kind;
+            var right = Language.FlipBracket(left);
+            var start = CurrentStart;
+            AcceptAndMoveNext();
+            if (EndOfFile && ((mode & BalancingModes.NoErrorOnFailure) != BalancingModes.NoErrorOnFailure))
             {
-                yield return CurrentToken;
-                NextToken();
+                Context.ErrorSink.OnError(
+                    RazorDiagnosticFactory.CreateParsing_ExpectedCloseBracketBeforeEOF(
+                        new SourceSpan(start, contentLength: 1 /* { OR } */),
+                        Language.GetSample(left),
+                        Language.GetSample(right)));
             }
+
+            return Balance(builder, mode, left, right, start);
         }
 
-        private void Configure(SpanKindInternal? kind, AcceptedCharactersInternal? accepts)
+        protected internal bool Balance(SyntaxListBuilder<RazorSyntaxNode> builder, BalancingModes mode, SyntaxKind left, SyntaxKind right, SourceLocation start)
```

Comment:
I believe we only ever balance in C#, if that's the case all of our balance methods should go in the csharp code parser.